v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/JsonStringToMutationFn.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.transforms;
+
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.Mutation.WriteBuilder;
+import com.google.cloud.spanner.Value;
+import com.google.cloud.teleport.v2.utils.SpannerUtils;
+import com.google.cloud.teleport.v2.values.SpannerSchema;
+import com.google.cloud.teleport.v2.values.SpannerSchema.SpannerDataTypes;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.json.JSONObject;
+
+/** Cloud Spanner transformation helper that convert each row to Mutation object. */
+public class JsonStringToMutationFn extends DoFn<String, Mutation> {
+
+  private final String projectId;
+  private final String spannerTableName;
+  private final String spannerDBName;
+  private final String spannerInstanceName;
+
+  private SpannerSchema spannerSchema = null;
+  private List<String> spannerSchemaColumns = null;
+
+  public JsonStringToMutationFn(
+      String projectId, String spannerInstanceName, String spannerDBName, String spannerTableName) {
+    this.spannerTableName = spannerTableName;
+    this.spannerDBName = spannerDBName;
+    this.spannerInstanceName = spannerInstanceName;
+    this.projectId = projectId;
+  }
+
+  @Setup
+  public void setup() {
+    this.spannerSchema =
+        SpannerUtils.getSpannerSchemaFromSpanner(
+            this.projectId, spannerInstanceName, spannerDBName, spannerTableName);
+    this.spannerSchemaColumns = spannerSchema.getColumnList();
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) throws IOException {
+    JsonObject row = JsonParser.parseString(Objects.requireNonNull(c.element())).getAsJsonObject();
+    WriteBuilder writeBuilder = Mutation.newInsertOrUpdateBuilder(spannerTableName);
+    c.output(parseRow(writeBuilder, row, spannerSchema));
+  }
+
+  /**
+   * Take a String list and convert into a Mutation object base on schema provided.
+   *
+   * @param builder MutationBuilder to construct
+   * @param row CSVRecord parsed list of data cell
+   * @param spannerSchema list with column name and column data type
+   * @return build Mutation object for the row
+   */
```

Comment:
JavaDoc unclear, what's the string list and what it has to do with CSVRecord?

Reply:
Changed